Add support for TLSMode.ISTIO_MUTUAL when building cluster config.

incfly opened the pull request:
This is for #5059 clean up some behavior before we generate multi filter chains using envoy tls multiplexing.

Authn policy also stops inferring the cluster's tls settings, so the outcome is DR is the only truth where client side tls settings come from.

_Why in master_: plan of the record, if works fine, can be shipped into 0.8 or 0.8.1 by merging master into release-0.8 branch.

_Testing_:  Manual Testing TODO.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **geeknoid**

Assign the PR to them by writing `/assign @geeknoid` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["geeknoid"]} -->

incfly commented:
@rshriram FYI

With closer looking at the code, I realized we have to modify `cluster.go` instead of authn plugin. Because that's the place we build both default and subset clusters config. whereas plugin is only involved with default cluster settings.

rshriram commented:
Plugin is invoked for all clusters

rshriram commented:
The reason for requesting changes is because this can be done cleanly in the plugin by making sure it’s called for all clusters and encapsulating the service account stuff in the auth plugin.

incfly commented:
@rshriram you're right on this is called for all clusters including subset. https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L114

So the determining config field comes from DR (TLSSettings), the question is who should consume this settings. I agree with @diemtvu that the consumer of this field should be in clusters.go, so the authn only focused on the server authn requirement, and let networking side deal with client settings. This is the intended separate of client/server we want to have. Will be nice to make implementation aligned with user facing config.

If the concern is more of "why networking now needs to know where istio provided key cert live, etc?", we can extract those key certs paths logic and put it somewhere else.

@costinm  wdyt?

Regarding why in master, we plan to make it work E2E first and only after ensuring it works, then merge with release branch.

> Also does this mean we will no longer have a global auth setting and setup client side auth by “default”?

This is correct... We need to provide something global else after these work...

rshriram commented:
My point is retain the isDestinationEnabledForMtls. In that function, check global flag or local dest rule that comes as part of the plugin call. Add an extra parameter to plugin calls in cluster.go to supply the destination rule as well. 

This way, there is very minimal code change

istio-testing commented:
@incfly: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | aa8ca06e3e3d27b70b2a9daa32fa46f4655acfb1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6808/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | aa8ca06e3e3d27b70b2a9daa32fa46f4655acfb1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6808/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | aa8ca06e3e3d27b70b2a9daa32fa46f4655acfb1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/641/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rshriram commented on pilot/pkg/networking/plugin/authn/authentication.go:
```diff
@@ -334,47 +325,4 @@ func (Plugin) OnInboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.
 // OnOutboundCluster implements the Plugin interface method.
 func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service *model.Service,
 	servicePort *model.Port, cluster *xdsapi.Cluster) {
```
change this method signature to take the DestinationRule as well. Same goes for Inboundcluster. Then you can simply do if destRule != nil and destRule.trafficPolicy.TLSmode==istio_mutual then...
don't forget to extract the port specific traffic policy first.

rshriram replied:
cc @ostromart

costinm replied:
It seems most TLS related settings are in cluster.go - it may keep them grouped ( and maybe later move to a 'tls.go' plugin). The authentication plugin seems to deal mostly with the auth policies - while cluster with the network destination rules.

incfly replied:
@rshriram could we proceed as it is right now, given other TLSettings mode is also handled in cluster.go? And later if needed extract client tls stuff out into tls.go plugin, suggested by @costinm 

I'll send follow-up PR to actually build multi-filter chain + tls sniffing filter, which depends on this one to be checked-in.

Thanks!

rshriram replied:
Sure. But can you push this PR to 0.8 as well ? This is independent of any thing else.

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | aa8ca06e3e3d27b70b2a9daa32fa46f4655acfb1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/641/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6818/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6818/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/653/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rshriram commented:
Also this has nothing to do with tls sniffing. This pr is useful by itself in existing 0.8, where auth overrides tls settings in dest rule. So I still believe that 0.8 is the right place for the ISTIO_mutual stuff. The tls snuffing changes in inbound listener is a different PR

In auth plugin 
OnOutboundCluster(cluster, destination rule, service, port)
  If isIstioMtlsEnabled()
    Set istio mTLS

isIstioMtlsEnabled(mesh, destrule)
 if mesh.mtls and destrulr.tls == nil
    or  destrule.tlsmode == istio_mutual
   return true
 Return false

diemtvu commented:
Technically, the approach Shriram suggested could work. However, per discussion with Louis, Costin ([meeting note](https://docs.google.com/document/d/1CJoAF8GS7F6iYFGkt8WsvPgC4l1nnShFHAToJNSCQMA/edit)), we want to cleanly separate authn policy to affect server side only, and destination rule to affect client side only. As a result, I think it's make more sense to remove OutboundCluster authn plugin code, to the applyTrafficPolicy module.

We need to infer a 'default' destination rules from the global mesh config flag for now, for backward compatible. However, in the long run, I think we can remove that flag and have customers explicitly submit global authN policy and destination rule for that.

diemtvu commented on pilot/pkg/networking/core/v1alpha3/cluster.go:
```diff
@@ -359,6 +362,34 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
 		}
+	case networking.TLSSettings_ISTIO_MUTUAL:
+		// TODO(incfly): needs to change when SDS is ready.
+		cluster.TlsContext = &auth.UpstreamTlsContext{
```
Maybe change the authN plugin function to a regular function to set these config and call it here. With that, the code that related to Istio-mTLS still stays closed to each other.

rshriram commented on pilot/pkg/networking/core/v1alpha3/cluster.go:
```diff
@@ -89,7 +91,8 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(env model.Environmen
 			clusterName := model.BuildSubsetKey(model.TrafficDirectionOutbound, "", service.Hostname, port)
-			defaultCluster := buildDefaultCluster(env, clusterName, convertResolution(service.Resolution), hosts)
+			serviceAccounts := env.ServiceAccounts.GetIstioServiceAccounts(service.Hostname, []string{port.Name})
+			defaultCluster := buildDefaultCluster(env, clusterName, convertResolution(service.Resolution), hosts, serviceAccounts)
```
Alternate simpler implementation suggestion: when a destinationrule is obtained, pass it to a convertIstioMtlsToMtls function. This function will simply set the tls mode to mutual with the cert file names using istio certs and SAN field using the service accounts. Once converted, the rest of the code does not need to be changed at all. And if config is nil, create a new dest rule if mesh.mtls is true and set the mTLS settings in the dest rule. This way you don’t have to wire up tnt service accounts everywhere. 
In future, if anyone were to add more cluster building code, they wouldn’t have to worry about plumbing service accounts everywhere. If you want, you could add this code directly into model.DestinationRules() such that you don’t even have to change cluster.go

rshriram replied:
I suggest the conversion as it ensures that any other code using cluster.go functions does not end up overriding auth stuff.

rshriram replied:
this conversion is better done in cluster.go, instead of model code due to certain wildcard issues in destination rules. But please convert so that you don’t have to push the service account thing into all functions.

diemtvu replied:
It's a good suggestion. In fact, it's similar to authn policy: when there is no authn policy in CRD, we generate one based on config map settings. May be we can do the same here?

incfly commented:
@rshriram We also think this is a useful change for 0.8, regardless of tls multiplexing. I'll work on testing this against 0.8 branch and if we can get test passing before release cut, will change this into 0.8.

diemtvu commented:
This change will require customer to set up destination policy to make
authn policy work. I'm just worry that we don't have time to test the whole
thing through, given both DR and authN are relatively new in this release.
I think it's better to have this in 0.8.1.

On Tue, May 8, 2018 at 2:13 PM Jianfei Hu <notifications@github.com> wrote:

> @rshriram <https://github.com/rshriram> We also think this is a useful
> change for 0.8, regardless of tls multiplexing. I'll work on testing this
> against 0.8 branch and if we can get test passing before release cut, will
> change this into 0.8.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/5418#issuecomment-387544144>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AX99cYyoKcEdA8RZxEZZfYAJL3Z5wy2yks5twgpkgaJpZM4TzCmp>
> .
>


-- 

Diem Vu | Software Engineer | diemvu@google.com | +1 408-215-8127

diemtvu commented:
with v1/v1apha1, authN policy will set both client and server side, so it still work without destination rules. Do you think that is enough?

incfly commented:
I changed to fill key/cert path in advance, instead of plumbing service account everywhere. I still want to keep ISTIO_MUTUAL separate from MUTUAL because in future when SDS is available we can't use key/certs paths to configure envoy.

Also removed the change on buildDefaultCluster, which is used for user app container as cluster as well. And there, we shouldn't add any TLSSettings.

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6818/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/653/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | ac54079b9eb37872e374687d8371f807c15b6187 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7074/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6818/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/653/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7075/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rshriram commented on pilot/pkg/networking/core/v1alpha3/cluster.go:
```diff
@@ -89,7 +91,14 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(env model.Environmen
+
+			// set TLSSettings if configmap global settings specifies MUTUAL_TLS.
+			if env.Mesh.AuthPolicy == meshconfig.MeshConfig_MUTUAL_TLS {
+				applyUpstreamTLSSettings(defaultCluster, buildIstioMutualTls(upstreamServiceAccounts))
```
Bug: this cannot be applied for external services. Add a check for service.MeshExternal and skip setting up this istio mutual for default

rshriram replied:
I also don’t know how this will work for headless services (ie ones using original dst cluster). You have to enable the auth in headless service test and see. If it fails add a check for original dst cluster here as well, just like the plugin code

incfly replied:
I checked external service now. Sorry for the lack of familiarity, what's the headless services/original dst cluster? is there some field to check, similar as `service.MeshExternal`?

rshriram commented on pilot/pkg/networking/core/v1alpha3/cluster.go:
```diff
@@ -214,6 +224,30 @@ func convertResolution(resolution model.Resolution) v2.Cluster_DiscoveryType {
+		if subset.TrafficPolicy.Tls.Mode == networking.TLSSettings_ISTIO_MUTUAL {
+			subset.TrafficPolicy.Tls = buildIstioMutualTls(upstreamServiceAccount)
+		}
+	}
```
This is not enough. We also have port traffic policy - which is port specific traffic policy. See code in this file for example b

incfly replied:
Done.

However, this concerns me that we potentially use `TrafficPolicy.TLSSettings` in lots of places of `DestinationRule`, and it's cumbersome to keep the conversion in sync while we're adding other level (label, say), traffic policy override. Unless we do proto reflection to traverse over the DestinationRule and find all fields typed as TLSSettings...

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 064753cba14d4d2fc22b4d7144775c2bb223629b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6818/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7075/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/790/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1334/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7075/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/790/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1334/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6954/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/790/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1334/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6954/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7084/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

incfly commented on pilot/pkg/networking/core/v1alpha3/cluster.go:
```diff
@@ -89,7 +91,14 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(env model.Environmen
 			defaultCluster := buildDefaultCluster(env, clusterName, convertResolution(service.Resolution), hosts)
+
+			// set TLSSettings if configmap global settings specifies MUTUAL_TLS.
+			if env.Mesh.AuthPolicy == meshconfig.MeshConfig_MUTUAL_TLS && !service.External() {
```
Yeah, I found that and changed to MeshExternal in the latest commit.

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1334/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6954/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7084/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/797/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1334/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6954/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/797/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7086/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | d39ad424fb4ccd0ed5f1f1cdafb1a117cc3204da | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6954/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/797/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7086/) | `/test istio-unit-tests`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1341/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/797/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7086/) | `/test istio-unit-tests`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1341/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6961/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7086/) | `/test istio-unit-tests`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1341/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | 3bc5fe2e65839ffa88bf05d34053cf0ac0bf6e5d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6961/) | `/test istio-pilot-e2e`
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/800/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7086/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/800/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1344/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6964/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/800/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1344/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6964/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7099/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1344/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6964/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7099/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/805/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6964/) | `/test istio-pilot-e2e`
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7099/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/805/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1349/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-unit-tests/7099/) | `/test istio-unit-tests`
prow/e2e-bookInfoTests-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/e2e-bookInfoTests-envoyv2-v1alpha3/805/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e-envoyv2-v1alpha3/1349/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | ca83544446e16213646008212e809d680b4aa0b5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/5418/istio-pilot-e2e/6969/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@incfly: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
